File: code/Python/Chapter4/4-2/bracket.py
def is_valid_bracket(bracket):
    stack = []

    for c in bracket:
        if c == "(" or c == "[" or c == "{":
            stack.append(c)
        else:
            if len(stack) == 0:
                return False
            else:
                top = stack.pop()
                if (top == "(" and c != ")") or (top == "[" and c != "]") or (top == "{" and c != "}"):
                    return False
    return len(stack) == 0

brackets = [
    "()",
    "()[]{}",
    "(]",
    "([)]",
    "{[]}",
]

for bracket in brackets:
    print(bracket, ":", is_valid_bracket(bracket))

File: code/C++/Chapter4/4-2/bracket.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <iomanip>

using namespace std;

bool is_valid_bracket(string bracket) {
    stack<char> stack;

    for (char c : bracket) {
        if (c == '(' || c == '[' || c == '{') {
            stack.push(c);
        } else {
            if (stack.empty()) {
                return false;
            }

            char top = stack.top();
            stack.pop();

            if ((c == ')' && top != '(') || (c == ']' && top != '[') || (c == '}' && top != '{')) {
                return false;
            }
        }
    }

    return stack.empty();
}

int main() {
    vector<string> brackets = {
        "()",
        "()[]{}",
        "(]",
        "([)]",
        "{[]}",
    };

    for (string bracket : brackets) {
        cout << bracket << ": " << boolalpha << is_valid_bracket(bracket) << endl;
    }

    return 0;
}